### DISCUSSION TITLE
Exception not being caught in try block

### DISCUSSION THREAD

**Discussion by osmanlodzero** [jchristn/WatsonTcp#171] [Q&A]
Hey, so I've been away from C# for a while so perhaps this is me overlooking something very obvious, but I've got this exception that's not getting caught in a try block when I try to access null Metadata. I did some googling and some suggest it could be the setting in the debugger called Just My Code but this was not the case after testing it out.

Others suggested it could be related to async so I thought I'd post here asking if anyone knows what this could be, thanks in advance!
![Exception not caught ](https://i.imgur.com/J0EWMzN.png)

**Comment by jchristn** [comment]
If ```Metadata``` is set to ```null``` when sending the message, then it would be ```null``` on the receiving side.  This line of code is attempting to retrieve the key ```Type``` from a null object.  

Are you sure the other side is sending metadata?  A better way to go about this would be to do something along these lines:
```
if (Message.Metadata != null && Message.Metadata.ContainsKey("Type"))
{
  string MsgType = (string)Message.Metadata["Type"];
}
```

**Reply by osmanlodzero** [reply]
Hey @jchristn , in the test above the metadata was deliberately null and my actual code now does indeed check for null and key.

What I was curious about in my original question was: why does the code that I've posted not catch the NullReferenceException but it rather breaks. Shouldn't it have gone into the catch block?

Thanks!

**Reply by jchristn** [reply]
Ah, got you.  Interesting question.  What happens if you change ```catch (System.Exception)``` to just ```catch (Exception)```?  I'm wondering if by specifying the namespace it's expecting a generic ```Exception``` as compared to something that inherits ```Exception```?  I really don't know, honestly.

**Reply by osmanlodzero** [reply]
Yeah I usually never specify a namespace with exceptions, but after a google search someone suggested it and it was there when I took a screenshot. But yes, even without it, it breaks, which has puzzled me.

**Reply by jchristn** [reply]
Great question @osmanlodzero I'll leave it open to the community; I don't know the answer.
